File: LeetCode/Medium/0086-partition-list/0086-partition-list.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func partition(head *ListNode, x int) *ListNode {
    smallDummy := &ListNode{}
    bigDummy := &ListNode{}
    smallDummyHead := smallDummy
    bigDummyHead := bigDummy
    
    for head != nil {
        if head.Val < x {
            smallDummy.Next = head
            smallDummy = head
        } else {
            bigDummy.Next = head
            bigDummy = head
        }
        head = head.Next
    }
    
    smallDummy.Next = bigDummyHead.Next
    bigDummy.Next = nil
    
    return smallDummyHead.Next
}

File: LeetCode/Medium/0086-partition-list/0086-partition-list.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode partition(ListNode head, int x) {
        ListNode smallDummy = new ListNode();
        ListNode bigDummy = new ListNode();
        ListNode smallDummyHead = smallDummy;
        ListNode bigDummyHead = bigDummy;
        
        while (head != null){
            if (head.val < x){
                smallDummy.next = head;
                smallDummy = head;
            } else {
                bigDummy.next = head;
                bigDummy = head;
            }
            head = head.next;
        }
        
        smallDummy.next = bigDummyHead.next;
        bigDummy.next = null;
        
        return smallDummyHead.next;
    }
}